Linux: Nuke Stat bits in favour of Statx

The-King-of-Toasters opened the pull request:
Maintaining the POSIX `stat` bits for Zig is a pain. Not only is `struct stat` incompatable between architectures, but maddingly annoying so; timestamps are specified as machine longs or fixed-width ints, members can be signed or unsigned. The libcs deal with this by introducing the own version of `struct stat` and copying the kernel structure members to
it. In the case of glibc, they did it twice thanks to the largefile transition!

In practice, the project needs to maintain three versions of `struct stat`:
- What the kernel defines.
- What musl wants for `struct stat`.
- What glibc wants for `struct stat64`. Make sure to use `fstatat64`!

And it's not as simple as running `zig translate-c`. In #21440 I had to create test programs in C and use `pahole` to dump the structure of `stat` for each arch, and I was constantly running into issue regarding padding and signed/unsigned ints. The fact that so many target checks in the `linux` and `posix` tests exist is most likely due to writing to padding bits and failing later.

The solution to this madness is `statx(2)`:
- It takes a single structure that is the same for all arches AND libcs.
- It uses a custom timestamp format, but it is 64-bit ready.
- It gives the same info as `fstatat(2)` and more!
- Unlike `fstatat(2)`, you can request a subset of the info required based on passing a mask.

It's so good that modern Linux arches (e.g. riscv) don't even implement `stat`, with the libcs using a generic `struct stat` and copying from `struct statx`.

Therefore, this PR rips out all the `stat` bits from `std.os.linux` and `std.c`. `std.posix.Stat` is now `void`, and calling `std.posix.*stat` is an compile-time error. A wrapper around `statx` has been added to `std.os.linux`, and callers have been upgraded to use it. Tests have also been updated to use `statx` where possible.

While I was here, I converted the mask and file attributes to be packed struct bitfields. A nice side effect is checking that you actually received the members you asked for via `Statx.mask`, which I have used by adding `assert`s at specific callsites.

Progress towards #21738

IamSanjid committed:
Update std.zig.system.NativePaths.detect to support some more flags on NixOS like environment.

Basically detect `-idirafter` flag in `NIX_CFLAGS_COMPILE` and treat it like `-isystem`, also detect `NIX_CFLAGS_LINK` environment variable and treat it like the `NIX_LDFLAGS` .

Reference:
https://github.com/NixOS/nixpkgs/blob/74eefb4210ef545cebd6a7a25356734b0b49854f/pkgs/build-support/build-fhsenv-chroot/env.nix#L83

FnControlOption committed:
Return a `usize` from `@abs` if given an `isize`

Also:
- `c_ushort` for `c_short`
- `c_uint` for `c_int`
- `c_ulong` for `c_long`
- `c_ulonglong` for `c_longlong`

The-King-of-Toasters committed:
Linux: Update `Statx` structure

Also removes the blank lines between members, and a comptime sizeOf
check.

The-King-of-Toasters committed:
Linux: Nuke `Stat` bits in favour of statx

Maintaining the POSIX `stat` bits for Zig is a pain. The order and
bit-length of members differ between all architectures, and int types
can be signed or unsigned. The libcs deal with this by introducing the
own version of `struct stat` and copying the kernel structure members to
it. In the case of glibc, they did it twice thanks to the largefile
transition!

In practice, the project needs to maintain three versions of `struct
stat`:
- What the kernel defines.
- What musl wants for `struct stat`.
- What glibc wants for `struct stat64`. Make sure to use `fstatat64`!

This isn't as simple as running `zig translate-c`. In #21440 I had to:
- Compile toolchains for each arch+glibc/musl combo.
- Create a test `fstat` program with/without `FILE_OFFSET_BITS=64`.
- Dump the value for `struct stat`.
- Stare at `std.os.linux`/`std.c` and cry.
- Add some missing padding.

The fact that so many target checks in the `linux` and `posix` tests
exist is most likely due to writing to padding bits and failing later.

The solution to this madness is `statx(2)`:
- It takes a single structure that is the same for all arches AND libcs.
- It uses a custom timestamp format, but it is 64-bit ready.
- It gives the same info as `fstatat(2)` and more!
- Unlike `fstatat(2)`, you can request a subset of the info required
  based on passing a mask.

It's so good that modern Linux arches (e.g. riscv) don't even implement
`stat`, with the libcs using a generic `struct stat` and copying from
`struct statx`.

Therefore, this commit rips out all the `stat` bits from `std.os.linux`
and `std.c`. `std.posix.Stat` is now `void`, and calling
`std.posix.*stat` is an compile-time error. A wrapper around `statx` has
been added to `std.os.linux`, and callers have been upgraded to use it.
Tests have also been updated to use `statx` where possible.

While I was here, I converted the mask and file attributes to be packed
struct bitfields. A nice side effect is checking that you actually
recieved the members you asked for via `Statx.mask`, which I have used
by adding `assert`s at specific callsites.

The-King-of-Toasters committed:
Linux: Dedupe generic decls

Commit #fc7a5f2 moved many of the `_t` types up a level, but didn't
remove them from arch_bits. Since `Stat` is gone, all but `time_t` can
be removed.

rootbeer commented on lib/std/os/linux.zig:
```diff
@@ -10016,6 +10032,47 @@ pub const wrapped = struct {
+    };
+
+    pub fn statx(dirfd: fd_t, path: [*:0]const u8, flags: u32, mask: STATX) StatxError!Statx {
+        assert(mask.__RESERVED == false);
```
I don't think Zig needs to assert anything about `__RESERVED`.  That's the kernel's domain.

rootbeer commented on lib/std/posix/test.zig:
```diff
@@ -181,10 +182,24 @@ fn testReadlink(target_path: []const u8, symlink_path: []const u8) !void {
-test "linkat with different directories" {
-    if ((builtin.cpu.arch == .riscv32 or builtin.cpu.arch.isLoongArch()) and builtin.os.tag == .linux and !builtin.link_libc) return error.SkipZigTest; // No `fstatat()`.
-    if (builtin.cpu.arch.isMIPS64()) return error.SkipZigTest; // `nstat.nlink` assertion is failing with LLVM 20+ for unclear reasons.
+fn getLinkInfo(fd: posix.fd_t) !struct { posix.ino_t, posix.nlink_t } {
```
Nice!

rootbeer commented on lib/std/posix.zig:
```diff
@@ -116,7 +116,10 @@ pub const STDIN_FILENO = system.STDIN_FILENO;
 pub const SYS = system.SYS;
 pub const Sigaction = system.Sigaction;
-pub const Stat = system.Stat;
+pub const Stat = switch (native_os) {
```
I think mapping `Stat` to `void` deserves a comment (at least for the `.linux` prong).

The-King-of-Toasters commented:
Forgot to compile the compiler and added a call to `MappedFile`. Perhaps `fs.File.Stat` should add these members? I imagine it would be useful to know.

rootbeer reviewed (commented):
Looks reasonable to me.  I've got some questions, and ideas, but feel free to ignore them.

The-King-of-Toasters commented:
@rootbeer I'll take your comments and update the PR later today (Sydney time).
@alexrp How would you handle the failure in `glibc_runtime_check.zig`? Perhaps adding the `stat` function definitions but making the `struct stat` part a byte buffer instead.

alexrp commented:
Also, it doesn't have to be done in this PR, but it would be nice to decouple `std.c.dev_t` (& friends) completely from the corresponding types in `std.os.linux` since, for some of them, they're not the same size or even representation.

The-King-of-Toasters commented:
> Would you mind opening a mirror PR [on Codeberg](https://codeberg.org/ziglang/zig/pulls) for LoongArch and RISC-V CI?

[Done](https://codeberg.org/ziglang/zig/pulls/8).

bernardassan commented:
@The-King-of-Toasters could you adapt the structure I used in https://github.com/ziglang/zig/pull/25394 for the typed flags changes in `IoUring.zig` for `Statx.Attr` and `Statx.Mask` https://github.com/ziglang/zig/blob/321a739473b6ff2d3f97a4c251450be5ce17bfef/lib/std/os/linux.zig#L7064 and https://github.com/ziglang/zig/blob/321a739473b6ff2d3f97a4c251450be5ce17bfef/lib/std/os/linux.zig#L7122? In my opinion, that has better namespacing, and the doc comment allows users familiar with `STATX_ATTR_*` and `STATX_MASK_*` constants to easily find it when searching for them on https://ziglang.org/documentation/master/std/ which I think seems to be the main motivation for most of the identifiers been `UPPER_CASE` in `linux.zig`. We can do better in Zig :smile: than C, and this would also help to reduce conflicts when https://github.com/ziglang/zig/pull/25394 is been merged.

alexrp commented:
> * Should the statx wrapper exist?
>     * If so, should the structure be allocated & zeroed for you?

I think yes. We can always revisit that later, but let's stick with existing practice for now.

> * Should @bernardassan's naming style be preferred over the `UPPERCASE` way I have?

Seems fine I guess? But I don't have a strong opinion one way or the other.

The-King-of-Toasters commented:
Ahh, I see it got removed in 3bf0ce65a514e6f86241364c4a11089e32b3ba57. Guess I'll add a check against wasi nolibc.

mrdimidium committed:
std.posix.accept: handle non-listening socket EINVAL

jacobly0 committed:
x86_64: fix abi of a struct that partially fits in registers

Closes #26035

jedisct1 committed:
http.requestHasBody: a PUT response can have a body

HEAD/TRACE are bodyless, but PUT responses are body-capable
per RFC 7231.

jedisct1 committed:
std.crypto: add hybrid post-quantum/traditional key encapsulation

Hybrid KEMs combine a post-quantum secure KEM with a traditional
elliptic curve Diffie-Hellman key exchange.

The hybrid construction provides security against both classical and quantum
adversaries: even if one component is broken, the combined scheme remains
secure as long as the other component holds.

The implementation follows the IETF CFRG draft specification for concrete
hybrid KEMs:

https://datatracker.ietf.org/doc/draft-irtf-cfrg-concrete-hybrid-kems/

andrewrk committed:
std.Io.Threaded: fix the cancellation race

Now, before a syscall is entered, beginSyscall is called, which may
return error.Canceled. After syscall returns, whether error or success,
endSyscall is called. If the syscall returns EINTR then checkCancel is
called.

`cancelRequested` is removed from the std.Io VTable for now, with plans
to replace it with a more powerful API that allows protection against
cancellation requests.

closes #25751

Commit:
std.sort.partitionPoint: faster implementation (#30005)

Migrated from https://github.com/ziglang/zig/pull/21419

Co-authored-by: Jonathan Hallstrom <lmj.hallstrom@gmail.com>
Reviewed-on: https://codeberg.org/ziglang/zig/pulls/30005

alexrp committed:
ci: pin actions/checkout to current ziglang/checkout@HEAD

alexrp committed:
build: add -Dskip-spirv and -Dskip-wasm options

alexrp committed:
ci: skip spirv and wasm tests on x86_64-freebsd

These are already built and run on x86_64-linux.

Commit:
Merge pull request 'std.crypto: add hybrid post-quantum/traditional key encapsulation' (#30010) from jedisct1/zig:hybridkem into master

Reviewed-on: https://codeberg.org/ziglang/zig/pulls/30010

andrewrk committed:
issue templates: use anchors in URLs

alexrp committed:
README: github -> codeberg for some links

alexrp committed:
ci: apply workaround for #22213 to x86_64-linux scripts

alexrp committed:
Revert "ci: apply workaround for #22213 to x86_64-linux scripts"

This reverts commit a8f9b5dc0695bd0fb54d35f75c66175cc6ee87a8.

andrewrk committed:
std.Io.Threaded: tune requestCancel

On a heavily loaded Linux 6.17.5, I observed a maximum of 20 attempts
not acknowledged before the timeout (including exponential backoff) was
sufficient, despite the heavy load.

The time wasted here sleeping is mitigated by the fact that, later on,
the system will likely wait for the canceled task, causing it to
indefinitely yield until the canceled task finishes, and the task must
acknowledge the cancel before it proceeds to that point.

mlugg committed:
print_zir: fix typo

mlugg committed:
Sema: initialize OPV comptime allocs correctly

This was caused a `[0]std.builtin.Type.StructField.Attributes` to be
considered `undefined`, even though that type is OPV so should prefer
its OPV `.{}` over `undefined`.

Resolves: #30039

andrewrk committed:
std.Io.Threaded: use musl's beautiful pthread_cancel semantics

andrewrk committed:
std.os.linux: delete unnecessary `@compileError`

Without this, it already fails to compile with a sufficiently helpful
error message.

andrewrk committed:
std.Io.Threaded: fix compilation for riscv32-linux

andrewrk committed:
std.Io.Threaded: don't solve the cancel race after all

Unfortunately, trying again until the cancellation request is
acknowledged has been observed to incur a large amount of overhead,
and usually strong cancellation guarantees are not needed, so the
race condition is not handled here. Users who want to avoid this
have this menu of options instead:
* Use no libc, in which case Zig std lib can avoid the race (tracking
  issue: https://codeberg.org/ziglang/zig/issues/30049)
* Use musl libc
* Use `std.Io.Evented`. But this is not implemented yet. Tracked by
  - https://codeberg.org/ziglang/zig/issues/30050
  - https://codeberg.org/ziglang/zig/issues/30051

glibc + threaded is the only problematic combination.

andrewrk committed:
std.Io.Threaded: re-introduce retry logic behind config

alexrp committed:
issue templates: update issue labels to match the actual org labels

jedisct1 committed:
std.crypto.aes: expose the inverse MixColumns operation

The inverse MixColumns operation is already used internally for
AES decryption, but it wasn’t exposed in the public API because
it didn’t seem necessary at the time.

Since then, several new AES-based block ciphers and permutations
(such as Vistrutah and Areion) have been developed, and they require
this operation to be implementable in Zig.
Since then, new interesting AES-based block ciphers and permutations
(Vistrutah, Areion, etc). have been invented, and require that
operation to be implementable in Zig.

Commit:
Merge pull request 'Sema: initialize OPV comptime allocs correctly' (#30043) from reify-empty-struct into master

Reviewed-on: https://codeberg.org/ziglang/zig/pulls/30043

Rexicon226 committed:
link.Elf: implement aarch64 relocation

andrewrk committed:
std.Io.Threaded: make is_musl linux-only

andrewrk committed:
std.Io.Threaded: use the correct mmsghdr struct

Commit:
Merge pull request 'std.crypto.aes: expose the inverse MixColumns operation' (#30052) from jedisct1/zig:invmixcolumns into master

Reviewed-on: https://codeberg.org/ziglang/zig/pulls/30052

Commit:
std.zig.Ast: count addrspace tokens correctly

Before this PR this

```zig
const namespace = struct {
    extern const num: u8 addrspace(.generic);
};
// comment
```

got formatted to this

```zig
const namespace = struct {
    extern const num: u8 addrspace(.generic);
(
```

Co-authored-by: rpkak <rpkak@noreply.codeberg.org>
Co-committed-by: rpkak <rpkak@noreply.codeberg.org>

andrewrk committed:
std.os.linux: SIG enum is non-exhaustive

andrewrk committed:
std.Io.Threaded: musl: handle ECANCELED same as EINTR

Otherwise the pthread_cancel can affect unrelated tasks.

andrewrk committed:
std.Io.Threaded: don't use pthread_cancel with musl

It doesn't support setting the "canceled" status to false, so once a
thread has been canceled, all operations on the thread start permanently
failing.

andrewrk committed:
std.Io.Threaded: separate out ECANCELED handling again

If ECANCELED occurs, it's from pthread_cancel which will *permanently*
set that thread to be in a "canceling" state, which means the cancel
cannot be ignored. That means it cannot be retried, like EINTR. It must
be acknowledged.

AidanWelch committed:
std.Io.Timestamp: when creating a Clock.Timestamp actually set .raw instead of the non-existant .nanoseconds

lunagl committed:
Change github links to codeberg

Commit:
Merge pull request 'std.Io.Threaded: rework cancellation' (#30033) from cancellation into master

Reviewed-on: https://codeberg.org/ziglang/zig/pulls/30033

zihadmahiuddin committed:
std.process.ArgIteratorWasi: fix no-args deinit

jedisct1 committed:
Argon2: use the std.Io interface

Also reduce the memory required by tests.

4GB for every test is way too much and doesn't provide much benefits
in testing the algorithms.

alexrp committed:
ci: set a sensible maxrss in x86_64-windows scripts

alexrp committed:
build: adjust max_rss on a per-CI-host basis

This avoids pessimizing concurrency on all machines due to e.g. the macOS
machine having high memory usage across the board due to 16K page size.

This also adds max_rss to test-unit and test-c-abi since those tend to eat a
decent chunk of memory too.

alexrp committed:
std.process.Child: enable rusage collection for freebsd, illumos, serenity

alexrp committed:
std.c: add rusage for dragonfly, netbsd, openbsd

alexrp committed:
std.process.Child: enable rusage collection for dragonfly, netbsd, openbsd

alexrp committed:
build: add some missing darwin tags

alexrp committed:
ci: don't skip release mode tests on x86_64-windows-debug

pulsar is much faster than george so we don't need to do this anymore.

Commit:
Merge pull request 'Argon2: use the std.Io interface' (#30084) from jedisct1/zig:argon2 into master

Reviewed-on: https://codeberg.org/ziglang/zig/pulls/30084

alexrp committed:
ci: set maxrss from $ZSF_MAX_RSS if provided by the runner

All of our runners now define this. When running a CI script locally, this will
not be set, so we default to 0, aka "all available system memory".

mlugg committed:
std.heap: rework `c_allocator`

The main goal here was to avoid allocating padding and header space if
`malloc` already guarantees the alignment we need via `max_align_t`.
Previously, the compiler was using `std.heap.raw_c_allocator` as its GPA
in some cases depending on `std.c.max_align_t`, but that's pretty
fragile (it meant we had to encode our alignment requirements into
`src/main.zig`!). Perhaps more importantly, that solution is
unnecessarily restrictive: since Zig's `Allocator` API passes the
`Alignment` not only to `alloc`, but also to `free` etc, we are able to
use a different strategy depending on its value. So `c_allocator` can
simply compare the requested align to `Alignment.of(std.c.max_align_t)`,
and use a raw `malloc` call (no header needed!) if it will guarantee a
suitable alignment (which, in practice, will be true the vast majority
of the time).

So in short, this makes `std.heap.c_allocator` more memory efficient,
and probably removes any incentive to use `std.heap.raw_c_allocator`.

I also refactored the `c_allocator` implementation while doing this,
just to neaten things up a little.

alexrp committed:
ci: reduce x86_64-linux timeouts

These excessive timeouts should no longer be necessary with the recent tuning of
job capacity and maxrss on these machines.

kristoff-it committed:
make Io.net.sendMany compile

mlugg committed:
std.debug: skip `manage resources correctly` with cbe

alexrp committed:
build: bump test-zigc max_rss to 767_483_904 on x86_64-linux

alexrp commented:
https://codeberg.org/ziglang/zig/pulls/30122
